Improved version of the Class Register program

Agata-Gizinska opened the pull request:
This is the improved Class Register after implementing initial feedback from [the previous pull request](https://github.com/Agata-Gizinska/Class_register/pull/2). 

Moreover, new features has been implemented:
- add new student
- add new course
- give grade
- add student to existing course
- graduate student
- drop out student

Agata-Gizinska committed:
Introduced new feature Add new student; minor fixes in __init__ methods of Classroom and Course classes; added several descriptive comments

Agata-Gizinska committed:
Introduced Add new course feature; Fixed bug that prevented is_course_in_register method from finding existing course which is not in the first row of the register; minor unification of naming variables regarding the number of partial grades required to pass a course; minor change in output message if classroom or course don’t exist while creating new student

Agata-Gizinska committed:
Added read_rows_in_register method in Register and replaced ‘with open(file)’ statements in other class methods; added extract_classroom_info method in ClassroomRegister; minor changes and additions in placeholder methods; added is_student_in_register, is_student_attending_course and extract_student_info methods in StudentRegister; added extract_course_info in CourseRegister; changes in Student __init__ method: check if student object is already registered; minor formatting changes in ‘—new_student’ and ‘—new_course’ parser options; initial version of ‘—give_grade’ parser option

Agata-Gizinska committed:
docs: Changed some comments into docstrings and added some new docstrings throughout the program; Added some placeholder methods for dropping out student

Agata-Gizinska committed:
Fully functional '--give_grade' option

feat: Implemented update_student_info method of StudentRegister class;
Implemented move_student_to_graduates method of CourseRegister class;

fix: Changed csv.reader and csv.writer delimiter to semicolon to omit problems with lists in entries;
Updated extract_student_info method of StudentRegister class so it now handles data about courses properly;
Reformatted new_student method of StudentRegister class to csv.writer;
Updated Student class __init__ method so self.courses attribute is assigned properly despite the number of courses;
Updated get_grade method of Student class depending on the number of grades that student has already received for a course;
Updated ‘—new_student’ option to check if student is already in register;
Minor fixes in variables’ naming;
Minor fixes in string formatting throughout whole program;
Changed __repr__ methods in Classroom class;
Updated calc_final_grade method of Student class that strings are converted to integers;
Updated pass_course method of Course class;

docs: Minor changes in docstrings;
Added more descriptive comments.

Agata-Gizinska committed:
feat: enabled appending of an existing student to an existing course by adding ‘—append_to_course’ parser argument

fix:
- merged methods move_student_to_graduates and move_student_to_dropouts into change_student_status in ClassroomRegister class and in CourseRegister class;
- reformatted update_student_info method in StudentRegister class so it can be used when giving a grade to a student and adding a student to existing course;
- minor fix in add_student_to_course method of CourseRegister class;
- changed the sequence of methods in Classroom class;
- minor fix in __init__ method of Student class;
- updated add_to_course method in Student class;

Agata-Gizinska committed:
Bug fixes and started working on graduation feature

feat: began working on graduation feature:
- Implemented check_passed_courses method in Student class;
- Added new conditional statement in get_grade method in Student class;
fix:
- fixed bug in extract_student_info method in StudentRegister class;
- fixed bug in update_student_info method in StudentRegister class;
- fixed bugs in __init__ method in Student class;
- minor fix in add_to_course method in Student class;
- minor fix in add_student_to_course method in Course class.

Agata-Gizinska committed:
Graduation feature is operational

feat: implemented graduation feature;
fix:
- fixed sequencing of data in entries in change_student_status method in ClassroomRegister class;
- fixed bug in graduate_student method in Classroom class;
- fixed bug in graduate method in Student method;
- fixed bug in pass_course method in Course class;
docs:
- added informative print message that student has graduated

Agata-Gizinska committed:
Implemented drop out student feature

feat:
- implemented dropping out student feature
- implemented student's status in the Student Register
fix:
- grammar fixes in documentation
- improved naming of variables
- added student's status check to prevent giving grades to graduates and dropouts
docs:
- implemeted docstrings for classes
- improved docstrings for methods
- improved comments

czarnadzuma11 commented on class_register.py:
```diff
@@ -259,159 +680,464 @@ def __init__(self, first_name, last_name, birth_date, classroom, course,
+        checks if graduation conditions are met.
+        Takes arguments: student's fullname, course name, grade
+    """
+
```
Remove this, parser help already includes this information,

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -259,159 +680,464 @@ def __init__(self, first_name, last_name, birth_date, classroom, course,
-    # print Classroom Register
-    parser.add_argument('--print_class', action='store_true',
+    # Print the Classroom Register
+    parser.add_argument('--print_classes', action='store_true',
```
Explanation is already included in help=, remove comments

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -259,159 +680,464 @@ def __init__(self, first_name, last_name, birth_date, classroom, course,
+            new_classroom = Classroom(start_year, end_year)
+            class_reg.new_classroom(new_classroom)
+            print(f'{new_classroom} has been added to register')
+
```
Make those long elif cases into functions

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -19,188 +35,531 @@ def __init__(self):
+
+    Methods:
+    -------------
+    is_classroom_in_register(classroom) \n
```
Triple " will preserve the formatting, you do not need to add \n manually

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -19,188 +35,531 @@ def __init__(self):
-                    return classroom
-                else:
-                    return False
+    def is_classroom_in_register(self, classroom):
```
if function name begins with is_ I'd expect it to return true/false. You can rename it  to get_classroom and it will behave as the name would suggest

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -19,188 +35,531 @@ def __init__(self):
-                    return False
+    def is_classroom_in_register(self, classroom):
+        """Check if a classroom is in the Classroom Register."""
+        rows = self.read_rows_in_register()
```
Why would we not do self.read_rows_in_register() on register __init__ and at the same time do all of those strip and replaces that we need to do below?

Agata-Gizinska replied:
implemented

Agata-Gizinska committed:
Fixes and reformatting

fix:
- reformatted long if-elif into functions,
- implemented subparsers for each main feature,
- removed redundant docstrings and comments

Agata-Gizinska committed:
Implemented subparsers into main function

fix:
- reformatted argparse main function by introducing subparsers for each feature;
- reformatted child Register classes that string formatting is handled in the parent Register class;
- changed csv.reader and csv.writer into csv.DictReader and csv.DictWriter when csv files are opened in reading or reading+writing mode, adjusted methods to handle dictionaries properly;
- renamed methods in ClassroomRegister to get_classroom_from_register;
- renamed methods in CourseRegister to get_course_from_register;
- added update_student_info method into graduate and drop_out methods in Student class;
- implemented a condition in append_to_course_func, preventing from adding the student to more courses than are necessary to graduate.
docs:
- adjusted docstrings to changes in classes;
- added docstrings to new functions;
- added help message to ‘print_register’ subparser

Agata-Gizinska committed:
Implemented subparsers into main function

fix:
- reformatted argparse main function by introducting subparsers for each feature;
- reformatted child Register classes that string formatting is handled in the parent Register class;
- changed csv.reader and csv.writer into csv.DictReader and csv.DictWriter when csv files are opened in reading or reading+writing mode, adjusted methods to handle dictionaries properly;
- renamed methods in ClassroomRegister to get_classroom_from_register;
- renamed methods in CourseRegister to get_course_from_register;
- added update_student_info method into graduate and drop_out medthods in Student class;
- implemented a condition in append_to_course_func, preventing from adding the student to more courses than are necessary to graduate.
docs:
- adjusted docstrings to changes in classes;
- added docstrings to new functions;
- added help message to ‘print_register’ subparser

Agata-Gizinska committed:
Update README.md

Agata-Gizinska committed:
Merge remote-tracking branch 'origin' into devwork_new_features
